File: 1_Einfuehrung/MATLAB/MSMP_abewertung_Loesung.m
%% Frequenzbewertung
% In der Akustik werden in einigen Fällen die empfangenen Signale
% frequenzabhängig gewichtet, wobei die Gewichtung den Frequenzgang des
% menschnlichen Gehörs berücksichtigt. Das A-Bewertungsfilter ist im
% Bereich von 20 Hz bis 20 kHz definiert. Stellen Sie den Frequenzgang des
% A-bewerteten Filters im Bereich von 10 Hz bis 100 kHz grafisch dar, wobei
% der Definitionsbereich des Filters von 20 Hz bis 20 kHz beträgt.
%
% Messtechnische Signalanalyse mit MATLAB und Python, Vorlesung/Übung
% 
% (c) Elektrische Messtechnik, Universität Paderborn - http://emt.upb.de

%% Initialisierung
close all; clear all; clc;

%% Variablendefinition
f = logspace(log10(20), log10(20e3), 500);

Ra = 12194^2.*f.^4./((f.^2+20.6.^2) .* (f.^2+12194.^2) .* sqrt(f.^2+107.7^2) .* sqrt(f.^2+737.9.^2));
A = 20*log10(Ra) + 2;

%% Grafische Ausgabe
semilogx(f, A, 'LineWidth', 1.5)
hold on
semilogx([f(1) f(end)], [A(1) A(end)], 'or', 'LineWidth', 1.5)
grid on
xlabel('Frequenz {\itf} / Hz')
ylabel('Frequenzgang |{\itG}| / dB')
title('A-Bewertungsfilter')

File: 1_Einfuehrung/Python/MSMP_A_Bewertung_Loesung.py
'''
Frequenzbewertung
In der Akustik werden in einigen Fällen die empfangenen Signale
frequenzabhängig gewichtet, wobei die Gewichtung den Frequenzgang des
menschnlichen Gehörs berücksichtigt. Das A-Bewertungsfilter ist im
Bereich von 20 Hz bis 20 kHz definiert. Stellen Sie den Frequenzgang des
A-bewerteten Filters im Bereich von 10 Hz bis 100 kHz grafisch dar, wobei
der Definitionsbereich des Filters von 20 Hz bis 20 kHz beträgt.

Messtechnische Signalanalyse mit MATLAB und Python, Vorlesung/Übung

(c) Elektrische Messtechnik, Universität Paderborn - http://emt.upb.de
'''

import numpy as np
import matplotlib.pyplot as plt

# Variablendefinition
a, b, c, d, e = 12194, 20.6, 107.7, 737.9, 12194  # in Hz
f = np.logspace(np.log10(20), np.log10(20e3), 500)

Ra = a**2*f**4./((f**2+b**2) * np.sqrt(f**2+c**2) * np.sqrt(f**2+d**2) * (f**2+e**2))
A = 20 * np.log10(Ra) + 2

# Grafische Augabe
plt.figure(1)
plt.semilogx(f, A, linewidth=1.5)
plt.semilogx(f[0], A[0], 'ro')
plt.semilogx(f[-1], A[-1], 'ro')
plt.grid(True)
plt.xlabel('Frequenz $f / \mathrm{Hz}$')
plt.ylabel('Frequenzgang $|G| / \mathrm{dB}$')
plt.title('Bewertungsfilter A')
plt.show()
